sklearn/neighbors/loop.py

```diff
@@ -0,0 +1,447 @@
+        in the training set). As LoOP originally does not deal with new data,
+        this method is kept private.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features), default=None
+            The query sample or samples to compute the Local Outlier
+            Probability w.r.t. to the training samples. If None, makes
+            prediction on the training data without considering them as
+            their own neighbors.
+
+        Returns
+        -------
+        is_inlier : array, shape (n_samples,)
+            Returns -1 for anomalies/outliers and +1 for inliers.
+        """
+        check_is_fitted(self,
+                        ["threshold_", "negative_local_outlier_probability_",
+                         "n_neighbors_", "_distances_fit_X_"])
+
+        is_inlier = LocalOutlierMixin._score_samples(self, X, mode='loop')
+
+        return is_inlier
+
+    @staticmethod
+    def _ssd(distances_X):
+        """
+        Calculates the sum of square distance of X. As this method is used in
+        the calculation of the local probability of each sample, this method
+        is kept private.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            The pairwise distances between each sample in the input data.
+
+        Returns
+        -------
+        ssd_array : array, shape (n_samples,)
+            Returns the sum of square distance of X as an array of
+            repeated values.
+        """
+        ssd_array = np.sum(np.power(distances_X, 2), axis=1)
+
+        return ssd_array
+
+    def _standard_distances(self, ssd_X):
+        """
+        Calculates the standard distance of each sample in the input data. As
+        this method is used in the calculation of the local probability of each
+        sample, this method is kept private.
+
+        Parameters
+        ----------
+        ssd_X : array-like, shape (n_samples,)
+            The sum of square distance of X as an array of repeated values.
+
+        Returns
+        -------
+        standard_distances : array, shape (n_samples,)
+            The standard distance of each sample.
+        """
+        standard_distances = np.sqrt(np.divide(ssd_X, self.n_neighbors))
+
+        return standard_distances
+
+    def _prob_distances(self, standard_distances_X):
+        """
+        Calculates the probabilistic distance of each sample in the input data.
+        As this method is used in the calculation of the local probability of
+        each sample, this method is kept private.
+
+        Parameters
+        ----------
+        standard_distances_X : array-like, shape (n_samples,)
+            The standard distance of each sample.
+
+        Returns
+        -------
+        prob_distances : array, shape(n_samples,)
+            The probabilistic distances of each sample.
+        """
+        prob_distances = (self.norm_factor * standard_distances_X)
```

Comment:
The norm_factor on the probabilistic distances can be removed, as it will cancel out.
Hence the entire ssd/standard_distances/prob_distances construct is quite redundant, and can be streamlined into a single _prob_distances function.
C.f., Page 60 / 5.1.2 of Schubert, Erich (2013): Generalized and efficient outlier detection for spatial, temporal, and high-dimensional data mining.
https://edoc.ub.uni-muenchen.de/16693/1/Schubert_Erich.pdf